Clear and Search features

arl opened the pull request:
- added Clear Output Button
- added Search feature : 
  - highlighting of search results
  - loop through search result occurrences using next/prev buttons
  - search input background color to red if search string not found
  - shortcuts associated :
    - **Ctrl + F** -> show search box
    - **ESC** -> hide search box
    - **Return** when on search input -> next result occurrence
    - **SHIFT + Return** when on search input -> previous result occurrence

arl committed:
add search and clear buttons

arl committed:
first commit of 'search' feature

arl committed:
highlight search text

- when text change in search box, all corresponding strings are
highlighted
- next text change, we previsously cancel text highlighting before re-
applying using the modified text

arl committed:
add SearchWidget class and ui form

and minor modifications to mainwindow form

arl committed:
load searchwidget into the mainwindow

better without adding search to any layout...

arl committed:
load searchwidget at runtime

searchwindow:
	- no need to subclass QWidget, searchwidget.ui is loaded at
runtime
	- for this, searchwidget.ui file is integrated to a resource
file
	- remove all previous search related code from MainWindow

mainwindow:
	- re-add the vertical splitter

arl committed:
add searchhighliter class

arl committed:
animate search widget when shown/hidden

arl committed:
hide search widget on ESC keypress

arl committed:
search widget animation fix code and polish animation

there was probably a memory leak because the animation were created on
each show/hide of the search widget, and never deleted

clean code

arl committed:
loop through search results with prev/next buttons

- button signals directly connected to SearchHiglighter
- cyclically loop through search results occurences
- current occurence is highlited in another color

problem to correct : position of current occurence is lost if search
string is changed

arl committed:
return to focus after hiding search widget

return focus to previously focused widget whe search widget is hidden

arl committed:
search : record current occurence position

if search string is changed, search won't start from the beggining of
document. Instead it will start from the position of the previously
found occurence

arl committed:
ensure current search result is visible

arl committed:
align search widget to the right of mainoutput

arl committed:
rewrite event filter

arl committed:
search feature : improve UX

when search input line has focus :
- [Return] key acts like pressing 'next' button
- [SHIFT + Return] key acts like pressing 'prev' button

application-global keyboard shortcuts :
- [CTRL] + F make search widget appear
- [ESC] makes it disappear (made in another commit) only if search
	input has focus

arl committed:
set searchinput background to red when nothing found

arl committed:
clear buttons clear output text

arl committed:
also highlight search results in bottom output

without cursor

aurelien-rainone commented:
Quando avrò trovato perché il search input `QlineEdit` perde il focus quando dopo un click su prev/next button aggiungerò un commit...

arl committed:
fix comment typo

arl committed:
highlighter : use currentBlock().position() for exact text position

position returned was not always exact, inducing a slow shift. After
some thousands of chars, search results were not synchronized and thus
not showed on click prev/next
corrected with the use of currentBlock().position()

gvaldambrini commented on mainwindow.cpp:
```diff
@@ -142,9 +195,107 @@ void MainWindow::toggleOutputSplitter()
+    // to return focus to it when search widget is hidden
+    static QWidget *prevFocus = 0;
+
+    QPropertyAnimation *animation = new QPropertyAnimation(search_widget, "geometry");
```
Come già fatto nell'event filter, anche in questo caso non c'è bisogno di lavorare sul geometry, ma puoi impostare la proprietà "pos". Sbaglio?

arl replied:
Si hai ragione, mi serve solo la posizione, non toccò alla taglia dell' widget

arl committed:
search widget: animate pos, no need for geometry

gvaldambrini commented on mainwindow.cpp:
```diff
@@ -142,9 +195,107 @@ void MainWindow::toggleOutputSplitter()
+            {
+                // hide search widget on Escape key press
+                if (keyEvent->key() == Qt::Key_Escape)
+                    emit ui->searchButton->toggle();
```
http://doc.qt.io/qt-5/qabstractbutton.html#toggle
`toggle()` è uno slot, non un signal. Per una coincidenza (il fatto che emit sia espanso con una macro vuota) funziona lo stesso, ma è logicamente sbagliato.

gvaldambrini commented on mainwindow.cpp:
```diff
@@ -142,9 +195,107 @@ void MainWindow::toggleOutputSplitter()
+                    if (keyEvent->modifiers() == Qt::ShiftModifier)
+                        emit search_prev_button->click();
+                    else
+                        emit search_next_button->click();
```
Stessa cosa di sopra:
http://doc.qt.io/qt-5/qabstractbutton.html#click

arl committed:
remove unnecessary emit

gvaldambrini commented on searchhighlighter.cpp:
```diff
@@ -0,0 +1,148 @@
+    if (!_search_string.isEmpty())
+    {
+        if (!text.isEmpty())
+        {
```
Una piccola nota di stile: invece che fare un if dentro un if per includere tutto il codice della funzione, non potevi fare:

```
if (_search_string.isEmpty() || text.isEmpty())
    return;
```

? 
In questo modo, il corpo della funzione non sarebbe indentato, e ne guadagnerebbe la leggibilità.

gvaldambrini commented on searchhighlighter.h:
```diff
@@ -0,0 +1,110 @@
+    /// the current search string
+    QString _search_string;
+
+    /// indicate if highliter maintains an occurence cursor
```
typo: `highliter`

gvaldambrini commented on mainwindow.h:
```diff
@@ -31,10 +34,27 @@ class MainWindow : public QMainWindow
+     * - key : search text position
+     * - value : text char format before being highlighted
+     */
+    QMap<position_length_type, QTextCharFormat> search_results;
```
Non mi sembra sia usato in nessun modo, sbaglio?

arl replied:
Infatti, mi sono dimenticato di rimuovere queste linee, che usavo prima avere scoperto l'esistenza del `QSyntaxHighlighter`

gvaldambrini commented on mainwindow.h:
```diff
@@ -31,10 +34,27 @@ class MainWindow : public QMainWindow
+     * \brief pair {position ; length}
+     * (used for keeping track of highlighted search results)
+     */
+    typedef QPair<int, int> position_length_type;
```
Stesso discorso di search_results: direi che è inutilizzato e che quindi può essere rimosso.

gvaldambrini commented on searchwidget.cpp:
```diff
@@ -0,0 +1,28 @@
+      ui(new Ui::SearchWidget)
+{
+    ui->setupUi(this);
+    qDebug() << "SearchWidget::SearchWidget(" << parent << ")";
```
Questa qDebug non mi sembra aggiunga alcun valore, immagino sia stata aggiunta in una prima fase per vedere quando viene costruito il SearchWidget. Adesso possiamo rimuoverla?

gvaldambrini commented on searchwidget.cpp:
```diff
@@ -0,0 +1,28 @@
+#include <QDebug>
+
+SearchWidget::SearchWidget(QWidget *parent)
+    : QWidget(parent),
```
Non capisco questi file searchwidget.{h, cpp}.
Non sono inclusi nel progetto, e non mi sembra nemmeno che tu li stia utilizzando. Forse è una dimenticanza di quando non conoscevi l'`UiLoader`?

arl replied:
esatto

arl committed:
code style

arl committed:
correct comment typo

arl committed:
remove unused code and files

aurelien-rainone commented:
i nuovi commit contengono le modifiche relativi ai punti visti

gvaldambrini commented on mainwindow.cpp:
```diff
@@ -142,9 +195,105 @@ void MainWindow::toggleOutputSplitter()
+            }
+
+        }
+    }
```
Con queste modifiche all'eventFilter è entrata una regressione: il bottomOutput si può scrollare con il mouse.

arl committed:
fix regression with wheel in bottom output

gvaldambrini commented on searchhighlighter.cpp:
```diff
@@ -0,0 +1,145 @@
+        if (search_string_changed && _num_occurences > 0)
+        {
+            // last highlight loop, we didn't mark nothing because the cursor
+            // should be positionned before the previous position
```
typo: `positionned`

gvaldambrini commented on searchhighlighter.cpp:
```diff
@@ -0,0 +1,145 @@
+        return;
+
+    QTextCharFormat charFormat;
+    charFormat.setBackground(SEARCHRESULT_BACKCOL);
```
poiché lo setti ad ogni iterazione del while, non c'è bisogno di questa chiamata a setBackground

gvaldambrini commented on searchhighlighter.cpp:
```diff
@@ -0,0 +1,145 @@
+{
+    // reset all before highlight loop
+    _search_string = search;
+    _occurence_cursor = _num_occurences = occurence_pos = 0;
```
ho come l'impressione che `occurrence_pos` non serva a niente, visto che lo imposti sempre a 0 ma mai lo leggi. Sbaglio?

gvaldambrini commented on mainwindow.cpp:
```diff
@@ -142,9 +195,105 @@ void MainWindow::toggleOutputSplitter()
+
+void MainWindow::handleCursosPosChanged(int pos)
+{
+    // only in browsing mode
```
perché soltanto in "browsing mode"? E' vero che se arriva altro testo la posizione del cursore verrà automaticamente settata alla fine del documento, ma penso sia comunque utile (o perlomeno più intuitivo e non certo dannoso) che il documento possa essere scrollato in automatico per "seguire" il cursor della ricerca.

gvaldambrini commented on mainwindow.h:
```diff
@@ -13,15 +13,18 @@
 
 #include <QMainWindow>
+#include <QTextCharFormat>
+#include <QMap>
```
Questo include non dovrebbe più servirti

arl committed:
correct typo

arl committed:
remove duplicate call

arl committed:
remove unused variable

arl committed:
always scroll to show current search occurence

arl committed:
remove unusued include
